File: src/selmer/filter_parser.py
from __future__ import annotations

from typing import Any, Callable

from .safe import SAFE_CONTEXT_KEY, SafeValue, is_safe_value, safe, unwrap_safe
from .types import Context
from .util import (
    clj_str,
    get_accessor,
    get_in,
    is_escaping_variables,
    parse_accessor,
    should_filter_missing_values,
)


def escape_html_star(s: str) -> str:
    if not is_escaping_variables():
        return s
    return (
        s.replace("&", "&amp;")
        .replace("<", "&lt;")
        .replace(">", "&gt;")
        .replace('"', "&quot;")
        .replace("'", "&#39;")
    )


escape_html_bang = escape_html_star


def strip_doublequotes(s: str) -> str:
    return s[1:-1] if len(s) > 1 and s.startswith('"') and s.endswith('"') else s


def escape_html(x: Any) -> Any:
    unwrapped = unwrap_safe(x)
    if unwrapped is not x:
        return unwrapped
    return escape_html_star(clj_str(x))


def fix_filter_args(args: list[str]) -> list[str]:
    return [strip_doublequotes(arg) for arg in args]


def lookup_args(context: Context) -> Callable[[str], Any]:
    def lookup(arg: str) -> Any:
        if len(arg) > 1 and arg.startswith("@"):
            value = get_in(context, parse_accessor(arg[1:]))
            return arg if value is None else value
        return arg

    return lookup


def _split_respecting_quotes(s: str, delimiter: str) -> list[str]:
    result: list[str] = []
    buf = ""
    quoted = False
    escaped = False
    for ch in s:
        if escaped:
            buf += ch
            escaped = False
            continue
        if ch == "\\":
            buf += ch
            escaped = True
            continue
        if ch == '"':
            quoted = not quoted
            buf += ch
            continue
        if not quoted and ch == delimiter:
            result.append(buf)
            buf = ""
        else:
            buf += ch
    result.append(buf)
    return result


def filter_str_to_fn(s: str) -> Callable[[Any, Context], Any]:
    from .filters import get_filter

    parts = _split_respecting_quotes(s, ":")
    filter_name = parts[0].strip() if parts else ""
    args = fix_filter_args(parts[1:])
    filt = get_filter(filter_name)
    if filt is None:
        raise ValueError(f"No filter defined with the name '{filter_name}'")

    def run(x: Any, context: Context) -> Any:
        return filt(x, *[lookup_args(context)(arg) for arg in args])

    return run


def literal(value: str) -> bool:
    return (value.startswith('"') and value.endswith('"')) or value.isdigit()


literal_q = literal


def parse_literal(value: str) -> str:
    return value[1:-1] if value.startswith('"') else value


def split_value(s: str) -> list[str]:
    return [part.strip() for part in _split_respecting_quotes(s, "|") if part.strip()]


def _apply_filters(
    value: Any,
    body: str,
    filter_strings: list[str],
    compiled_filters: list[Callable[[Any, Context], Any]],
    context: Context,
) -> Any:
    acc = value
    for filter_string, filt in zip(filter_strings, compiled_filters, strict=False):
        try:
            acc = filt(acc, context)
        except Exception as exc:  # noqa: BLE001 - enrich template errors
            raise RuntimeError(
                f"On filter body '{body}' and filter '{filter_string}' this error occurred:{exc}"
            ) from exc
    return acc


def _finalize_filter_value(x: Any, context: Context, escape: bool) -> Any:
    unwrapped = unwrap_safe(x)
    if unwrapped is not x:
        return unwrapped
    if get_accessor(context, SAFE_CONTEXT_KEY):
        return unwrapped
    return escape_html(unwrapped) if escape else unwrapped


def get_accessor_value(m: Any, k: str | int) -> Any:
    return get_accessor(m, k)


def compile_filter_body(s: str, escape: bool = True) -> Callable[[Context], Any]:
    parts = split_value(s)
    val = parts[0] if parts else ""
    filter_strings = parts[1:]
    accessor = parse_accessor(val)
    compiled_filters = [filter_str_to_fn(filter_string) for filter_string in filter_strings]

    if literal(val):
        def run_literal(context: Context) -> Any:
            x = _apply_filters(parse_literal(val), s, filter_strings, compiled_filters, context)
            return _finalize_filter_value(x, context, escape)

        return run_literal

    def run(context: Context) -> Any:
        value = context
        for key in accessor:
            value = get_accessor(value, key)
        if value is not None or (should_filter_missing_values() and compiled_filters):
            x = _apply_filters(value, s, filter_strings, compiled_filters, context)
            return _finalize_filter_value(x, context, escape)
        return None

    return run

File: src/filter-parser.ts
import type { Context, FilterFunction } from "./types.js";
import { filters, getFilter } from "./filters.js";
import { SAFE_CONTEXT_KEY, safe, unwrapSafe, type SafeValue } from "./safe.js";
import {
  cljStr,
  getAccessor,
  getIn,
  isEscapingVariables,
  parseAccessor,
  shouldFilterMissingValues
} from "./util.js";

export { SAFE_CONTEXT_KEY, SAFE_FILTER, isSafeValue, safe, unwrapSafe, type SafeValue } from "./safe.js";

export function escapeHtmlStar(s: string): string {
  if (!isEscapingVariables()) return s;
  return s
    .replace(/&/g, "&amp;")
    .replace(/</g, "&lt;")
    .replace(/>/g, "&gt;")
    .replace(/"/g, "&quot;")
    .replace(/'/g, "&#39;");
}

export const escapeHtml$ = escapeHtmlStar;

export function stripDoublequotes(s: string): string {
  return s.length > 1 && s.startsWith('"') && s.endsWith('"') ? s.slice(1, -1) : s;
}

export function escapeHtml(x: unknown): unknown {
  const unwrapped = unwrapSafe(x);
  if (unwrapped !== x) return unwrapped;
  return escapeHtmlStar(cljStr(x));
}

export function fixFilterArgs(args: string[]): string[] {
  return args.map(stripDoublequotes);
}

export function lookupArgs(context: Context): (arg: string) => unknown {
  return (arg: string) => {
    if (arg.length > 1 && arg.startsWith("@")) {
      const value = getIn(context, parseAccessor(arg.slice(1)));
      return value === undefined || value === null ? arg : value;
    }
    return arg;
  };
}

function splitRespectingQuotes(s: string, delimiter: string): string[] {
  const result: string[] = [];
  let buf = "";
  let quoted = false;
  let escaped = false;
  for (const ch of s) {
    if (escaped) {
      buf += ch;
      escaped = false;
      continue;
    }
    if (ch === "\\") {
      buf += ch;
      escaped = true;
      continue;
    }
    if (ch === '"') {
      quoted = !quoted;
      buf += ch;
      continue;
    }
    if (!quoted && ch === delimiter) {
      result.push(buf);
      buf = "";
    } else {
      buf += ch;
    }
  }
  result.push(buf);
  return result;
}

export function filterStrToFn(s: string): (x: unknown, context: Context) => unknown {
  const [filterName = "", ...rawArgs] = splitRespectingQuotes(s, ":");
  const args = fixFilterArgs(rawArgs);
  const filter = getFilter(filterName.trim());
  if (filter) {
    return (x: unknown, context: Context) => filter(x, ...args.map(lookupArgs(context)));
  }
  throw new Error(`No filter defined with the name '${filterName.trim()}'`);
}

export function literal(value: string): boolean {
  return (value.startsWith('"') && value.endsWith('"')) || /^[0-9]+$/.test(value);
}

export const literal$ = literal;

export function parseLiteral(value: string): string {
  return value.startsWith('"') ? value.slice(1, -1) : value;
}

export function splitValue(s: string): string[] {
  return splitRespectingQuotes(s, "|").map((part) => part.trim()).filter((part) => part.length > 0);
}

function applyFilters(
  value: unknown,
  body: string,
  filterStrings: string[],
  compiledFilters: Array<(x: unknown, context: Context) => unknown>,
  context: Context
): unknown {
  return compiledFilters.reduce((acc, filter, index) => {
    const filterString = filterStrings[index];
    try {
      return filter(acc, context);
    } catch (cause) {
      const message = cause instanceof Error ? cause.message : String(cause);
      throw new Error(`On filter body '${body}' and filter '${filterString}' this error occurred:${message}`, {
        cause
      });
    }
  }, value);
}

export function compileFilterBody(s: string, escape = true): (context: Context) => unknown {
  const [val = "", ...filterStrings] = splitValue(s);
  const accessor = parseAccessor(val);
  const compiledFilters = filterStrings.map(filterStrToFn);

  if (literal(val)) {
    return (context: Context) => {
      const x = applyFilters(parseLiteral(val), s, filterStrings, compiledFilters, context);
      return finalizeFilterValue(x, context, escape);
    };
  }

  return (context: Context) => {
    const valFromContext = accessor.reduce((acc: unknown, key) => getAccessor(acc, key), context as unknown);
    if (valFromContext !== undefined && valFromContext !== null || (shouldFilterMissingValues() && compiledFilters.length > 0)) {
      const x = applyFilters(valFromContext, s, filterStrings, compiledFilters, context);
      return finalizeFilterValue(x, context, escape);
    }
    return undefined;
  };
}

function finalizeFilterValue(x: unknown, context: Context, escape: boolean): unknown {
  const unwrapped = unwrapSafe(x);
  if (unwrapped !== x) return unwrapped;
  if (getAccessor(context, SAFE_CONTEXT_KEY)) return unwrapped;
  if (escape) return escapeHtml(unwrapped);
  return unwrapped;
}

export { filters };
